### PR TITLE
Disallow function variants of in-place operators other than existing ones

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22720 Disallow function variants of in-place operators other than existing ones**

Related to #22707

### CODE REVIEW

**Comment by xuhdev** [comment]
Not sure whether the *elu functions in torch.nn require the same treatment (and they seem requiring special care for changing to method only). Should probably also add checks during the generation to ensure future compliance.

**Comment by xuhdev** [comment]
CI blocked by https://github.com/pytorch/vision/pull/1107

**Comment by gchanan** [comment]
does anyone (@suo?) know if this change could break serialization (i.e. do we serialize models in such a way that they would call the function variants?).

**Comment by gchanan** [comment]
Patch looks good.  I'd want to do a pass and figure out what this means for BC.

One quick thought: Just from a quick look, I don't see why we need copy_sparse_to_sparse_ exposed at all, it looks like implementations can just call it.

**Comment by xuhdev** [comment]
CI waiting for #22754

**Comment by xuhdev** [comment]
I also don't think copy_sparse_to_sparse_ should be exposed. But let's deal with that in another PR, because it looks like internal code depends on the fact that this function is generated from native_function.yaml

**Comment by gchanan** [comment]
I looked through some of these functions and they've existed for awhile, so there are BC concerns.  My take here is if we don't JIT serialize these, then I would add deprecation warnings to the methods (would involve codegen changes) in order to not just break our users for a minor benefit.

**Comment by xuhdev** [comment]
> I looked through some of these functions and they've existed for awhile, so there are BC concerns. My take here is if we don't JIT serialize these, then I would add deprecation warnings to the methods (would involve codegen changes) in order to not just break our users for a minor benefit.

Would you mind tagging #22707 or this as "need-discussion"? We can make a decision about how to deal with these functions before diving into the code...

**Comment by gchanan** [comment]
> Would you mind tagging #22707 or this as "need-discussion"? We can make a decision about how to deal with these functions before diving into the code...

done.

**Comment by xuhdev** [comment]
@gchanan This PR should be ready now. Feel free to review.

File: tools/autograd/gen_python_functions.py [hunk 1]
```diff
@@ -702,6 +706,17 @@ def process_function(name, declarations):
             typedef, next_index = emit_namedtuple_return_type_def(declaration, next_index)
             env['declare_namedtuple_return_types'] += typedef
 
+        # In-place operator, should be method only. Deprecate it.
+        if not is_module and not has_self and name.endswith('_') and not name.startswith('_'):
+            # After the deprecated functions are removed, replace this block with an AssertionError to ensure that
+            # future in-place operators won't be exposed as functions.
+            env['deprecation_message'] = ('PyErr_WarnEx(PyExc_DeprecationWarning, '
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: use TORCH_WARN instead.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -83,18 +83,22 @@
 - func: dropout(Tensor input, float p, bool train) -> Tensor
 
 - func: dropout_(Tensor(a!) self, float p, bool train) -> Tensor(a!)
+  variant: function, method
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: I'd prefer to have an indication in native_functions that something is wrong; it seems totally feasible that someone would copy-paste one of these, not happen to call the function variant and not get a warning.

I'm not sure if this would work because not all of these are deprecated, but I'd consider:
1) adding an annotation, e.g. "deprecated_method: True"
2) If you have one of these bad methods that don't have a deprecated_method tag (and isn't in NN?) you error out.

Is that feasible?

**Review by gchanan — COMMENTED** [review]
looks good!  One suggestion and one question.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (xuhdev) [reply aten/src/ATen/native/native_functions.yaml]: Do you think it would be simpler if we report an error in `gen_python_functions.py` (where TORCH_WARN is suggested), if the function is not one of the currently existing ones? I don't think anyone should add a new deprecated inplace operator that can be used as a function at this point (we are kind of inconsistent about whether an inplace operator has a function variant at this point).

And the time window to add a new such operator is probably short because we can actually come back and revisit this issue in two weeks, assuming a new release comes out before the end of July :)

Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: a whitelist also seems fine, as long as we still TORCH_WARN for those.

**Comment by xuhdev** [comment]
Do we also wanna do this for functions in the nn module in the future?

**Comment by gchanan** [comment]
I don't know, we don't really have a solution for that (I guess some functions use `inplace=True` but that seems weird...).

File: tools/autograd/gen_python_functions.py [hunk 3]
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
             typedef, next_index = emit_namedtuple_return_type_def(declaration, next_index)
             env['declare_namedtuple_return_types'] += typedef
 
+        # In-place operator, should be method only. Deprecate it.
+        if not is_module and not has_self and name.endswith('_') and not name.startswith('_'):
+            deprecated_functions = ['abs_', 'acos_', 'addmv_', 'alpha_dropout_', 'as_strided_', 'asin_', 'atan_',
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: ya, I don't think putting a big list here is a good idea.

For example, I don't see how we can deprecate a bunch of these -- our nn functional code calls a bunch of these, even if they aren't in the "correct" nn namespace (check all the dropouts, for example).

Comment (xuhdev) [reply tools/autograd/gen_python_functions.py]: I believe that the use of all of these function variants have been eliminated, at least in PyTorch codebase (No compilation error if we simply remove them from the scene).

Where would you like to put the list? This list is not supposed to grow anyway and we only hide them here temporarily.

What is your suggesetion for handling misplaced nn functions?

Comment (gchanan) [reply tools/autograd/gen_python_functions.py]: > I believe that the use of all of these function variants have been eliminated, at least in PyTorch codebase (No compilation error if we simply remove them from the scene).

What about something like this?
https://github.com/pytorch/pytorch/blob/eb76b7a564121c5fede749ad7d0a36f2b61a0a95/torch/nn/functional.py#L800

Comment (xuhdev) [reply tools/autograd/gen_python_functions.py]: OK, fair enough. I reverted all changes of all nn functions that are misplaced in `torch.` (in this context, `*dropout` functions). We can probably decide what we should do with those functions in the future. Could you review again?

Comment (xuhdev) [reply tools/autograd/gen_python_functions.py]: Do you in general treat `relu_` as an nn function? If so, I'll also remove its deprecation as well.

Comment (xuhdev) [reply tools/autograd/gen_python_functions.py]: @gchanan This PR should be ready now. Feel free to review. Thanks!

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please exclude `third_party/onnx` update

File: tools/autograd/gen_python_functions.py [hunk 4]
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
...
+                                    'embedding_renorm_', 'erf_', 'erfc_', 'exp_', 'expm1_', 'fill_', 'floor_', 'frac_',
+                                    'index_put_', 'log10_', 'log1p_', 'log2_', 'log_', 'neg_', 'reciprocal_', 'relu_',
+                                    'resize_as_', 'round_', 'rrelu_', 'rsqrt_', 'selu_', 'sigmoid_', 'sin_', 'sinh_',
+                                    'sqrt_', 'tan_', 'tanh_', 'threshold_', 'trunc_', 'zero_')
+            if name not in deprecated_functions:  # This is an in-place function variant that was later added
+                raise AssertionError('{} must not have a function variant.'.format(name))
+
+            # After the deprecated functions are removed, replace this block with an AssertionError to ensure that
+            # future in-place operators won't be exposed as functions.
+            env['deprecation_message'] = ('TORCH_WARN("In-place functions such as torch.{name} are deprecated '
```
Comment (VitalyFedyunin) [inline tools/autograd/gen_python_functions.py]: Please define template as global variable

File: tools/autograd/gen_python_functions.py [hunk 5]
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
             typedef, next_index = emit_namedtuple_return_type_def(declaration, next_index)
             env['declare_namedtuple_return_types'] += typedef
 
+        # In-place operator, should be method only. Deprecate it.
+        if not is_module and not has_self and name.endswith('_') and not name.startswith('_'):
+            deprecated_functions = ('dropout_', 'feature_dropout_', 'alpha_dropout_', 'feature_alpha_dropout_',
```
Comment (VitalyFedyunin) [inline tools/autograd/gen_python_functions.py]: Please move to global var and add comment explaining it's meaning

File: tools/autograd/gen_python_functions.py [hunk 6]
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
...
+            if name not in deprecated_functions:  # This is an in-place function variant that was later added
+                raise AssertionError('{} must not have a function variant.'.format(name))
+
+            # After the deprecated functions are removed, replace this block with an AssertionError to ensure that
+            # future in-place operators won't be exposed as functions.
+            env['deprecation_message'] = ('TORCH_WARN("In-place functions such as torch.{name} are deprecated '
+                                          'and will be removed in the next release. '
+                                          'Please use their in-place method counterparts (Tensor.{name}) instead.");'
+                                          ).format(name=name)
+            env['deprecated'] = True
```
Comment (VitalyFedyunin) [inline tools/autograd/gen_python_functions.py]: Unused

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]

File: tools/autograd/gen_python_functions.py [hunk 7]
```diff
@@ -66,6 +68,7 @@
 PY_VARIABLE_METHOD_NOARGS = CodeTemplate("""\
 static PyObject * ${pycname}(PyObject* self_, PyObject* args)
 {
+  ${deprecation_message}
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: hmm this is interesting, because removing these won't actually let us get rid of the methods, because this is only affecting the python API (and it currently must at least include this because the python functions can call C++ methods, so you can't put the deprecation warning in the C++ methods).

File: tools/autograd/gen_python_functions.py [hunk 8]
```diff
@@ -165,6 +168,20 @@
...
 
+INPLACE_DEPRECATION_MESSAGE = ('TORCH_WARN("In-place functions such as torch.{name} are deprecated '
+                               'and will be removed in the next release. '
+                               'Please use their in-place method counterparts (Tensor.{name}) instead.");')
+
+# These functions should not exist because they are in-place. We now deprecate them before removing them.
+DEPRECATED_INPLACE_FUNCTIONS = ('abs_', 'acos_', 'addmv_', 'as_strided_', 'asin_', 'atan_', 'ceil_', 'celu_',
+                                'clamp_', 'clamp_max_', 'clamp_min_', 'cos_', 'cosh_', 'detach_',
+                                'embedding_renorm_', 'erf_', 'erfc_', 'exp_', 'expm1_', 'fill_', 'floor_', 'frac_',
+                                'index_put_', 'log10_', 'log1p_', 'log2_', 'log_', 'neg_', 'reciprocal_', 'relu_',
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: how did you determine this list?  For example, it looks like rrelu_ is used in nn/functional, but it's not in the EXEMPTED list.

Did you check that none of these methods are actually used?

**Review by gchanan — CHANGES_REQUESTED** [review]
I'm pretty dubious about this change, because I think there are some pre-reqs we'd need to do first.

1) change python function bindings to call functions, not methods, and ensure there is no performance regression.
2) make sure we aren't introducing any warnings in our own code
3) warn at the C++ level, not just python.

File: tools/autograd/gen_python_functions.py (hunk 8 above)
Comment (xuhdev) [reply tools/autograd/gen_python_functions.py]: I'm still unsure yet what to do with `relu`, `rrelu`, and `celu`. These look like nn functions but they are under torch, and their positioning is kind of vague. For all others, I add them to deprecation list because they are in-place and semantically shouldn't be in nn.

**Comment by xuhdev** [comment]
> I'm pretty dubious about this change, because I think there are some pre-reqs we'd need to do first.
> 
>     1. change python function bindings to call functions, not methods, and ensure there is no performance regression.
> 
>     2. make sure we aren't introducing any warnings in our own code
> 
>     3. warn at the C++ level, not just python.

How about this: before we come back to think about how to deal with the deprecation, I'm gonna create another PR (or convert this PR) to stop additional inplace operators from being declared as function variants first (so as to prevent the situation from getting worse). Do you think it is sufficient to do this in tools/autograd/gen_python_functions.py by emitting an error if a new in-place operator is declared as function variant? It should look very similar to the current shape except that there is no deprecation message.

And Point 1 benefits not only this single deprecation purpose, but also the consistency within function and method variants.

**Comment by VitalyFedyunin** [comment]
> Do you think it is sufficient to do this in tools/autograd/gen_python_functions.py by emitting an error if a new in-place operator is declared as function variant? It should look very similar to the current shape except that there is no deprecation message.

Sounds good to me, as soon as we have detailed error message at codegen step.

**Comment by xuhdev** [comment]
@VitalyFedyunin  Updated

**Comment by VitalyFedyunin** [comment]
Hi! Can you please rebase.

**Comment by xuhdev** [comment]
@VitalyFedyunin Sorry for the delay. I've rebased now

**Comment by xuhdev** [comment]
@VitalyFedyunin Mind looking into this :)
